mitodl/mitxpro
# feat: Add support for dollars-off coupons

#### Pre-Flight checklist


#### What are the relevant tickets?
https://github.com/mitodl/mitxpro/issues/2443

#### What's this PR do?
- Adds support for discount type for dollars-off in Coupon Models
- Adds new field in `discount_type` CouponPaymentVersion
- Adds amount field validators based on discount type
- Refactors serializers and Django admin support for dollars-off coupon creation
- Updates coupon creation form to support different discount types like percent off and dollars off. Makes the relevant changes in the API payload and validations as well.
- Adds support for dollars-off coupons on the basket
- Refactors multiple checkout pages to support dollars-off discounts
- Adds dollars-off support for HubSpot
- Refactors and adds tests



#### How should this be manually tested?
This might need to be tested extensively. It might be a good idea to read the conversation on https://github.com/mitodl/mitxpro/issues/2443 before starting the testing.

- Visit Django Admin and check CouponPaymentVersion models and verify that you see `discount_type` field
- Visit `/ecommerce/admin/coupons/`
- Verify the discount type dropdown has options for percent-off and dollars-off.
- Verify the min and max values for the discount amount and relevant error messages.
- Verify that the correct amount and discount_type are sent in the payload. For percent-off amount is converted between 0.1-1.
- Create a couple of coupons 'dollars-off' or 'percent-off; either from Django Admin or through `http://xpro.odl.local:8053/admin/`
- Apply those coupons on an item in the basket (Make sure that all the cases are tested e.g. Global or course-based coupons work exactly as they should for both `dollars-off` and `percent-off`)
- Create dollars-off and percent-off coupons globally and test the case where we pick the best coupon automatically
- Check that the discounts and price displayed to the user is correct
- Disable coupons and check if they become invalid for both `dollars-off` and `percent-off`
- Check the Hubspot that the orders are created successfully there and also they contain the right discount and amounts as mentioned in https://github.com/mitodl/mitxpro/issues/2443
- For Hubspot test keys (You can ask me)
- Also check that the discount or price values never show up to be `negative` since we have dollars-off support now and with dollars, the values can be negative
- No `dollars-off` support for B2b coupons are only percent based and work as expected and coupons generated through b2b have `discount_type` to be `percent-off`
- No dollars-off support for coupons through sheets so they should work as expected and have `discount_type` to be `percent-off`.

#### Where should the reviewer start?
Coupons

#### Any background context you want to provide?

#### Screenshots (if appropriate)
<img width="1000" alt="Screenshot 2022-11-09 at 3 00 04 PM" src="https://user-images.githubusercontent.com/52656433/200801243-59af5f9d-24fd-4514-b0f9-a1adc048a272.png">

<img width="1000" alt="Screenshot 2022-11-09 at 3 00 11 PM" src="https://user-images.githubusercontent.com/52656433/200801277-f769be0b-d93b-4a78-a30a-8343e704c193.png">

<img width="1000" alt="Screenshot 2022-11-09 at 3 00 22 PM" src="https://user-images.githubusercontent.com/52656433/200801285-a91e846d-7f9b-4cb2-84f0-ae4bf420944c.png">

#### What GIF best describes this PR or how it makes you feel?
(Optional)
